### DISCUSSION TITLE
WAL performance

### DISCUSSION THREAD

**Discussion by potrusil-osi** [dotnet/dotNext#56] [General]
Hi sakno,

I've been testing performance of a distributed application built on top dotNext/Raft and ran into couple of issues with the current implementation of WAL. I'm using DotNext.AspNetCore.Cluster v3.0.2 and custom binary serialization through the command interpreter framework.

1. The biggest issue has been the compaction into a snapshot. When I configure everything optimally it takes around 100ms which is a huge toll on overall performance. The snapshot file is around 10MB and given the application I have, that's the lowest size I can get. In order to keep the compaction around 100ms it needs to be called often, and so the partition size is only 100 entries. I have two questions:
   1. Why is compaction running during AppendEntries message processing? Could it be moved into background?
   2. The buffer used for snapshot file creation is the same buffer that is used for writing entries into partition files. That doesn't seem to be optimal. I have observed that the ideal buffer size for the snapshot file is 1MB, while the ideal buffer size for the partition files is 15kB. Do you understand why there is such a difference? Could two different buffer sizes be supported?
2. The second issue occurs when on the leader node multiple threads try to append uncommitted entries concurrently (without any break). Understandably, there is a lock in PersistentState.AppendAsync. Since I'm testing with 30 client threads the average time to only lock the audit trail is around 40ms. Do you have any suggestions how to deal with such a problem?

**Comment by sakno** [comment]
Hi @potrusil-osi ,

Let's start from things that can be done easily:
1. How is your binary serialization implemented? If your log records structured with the fields of primitive types then methods such as `IAsyncBinaryWriter.WriteInt32Async` may slow down the performance. If the size of single log records is less than 1 KB then you can serialize the object into memory. `IAsyncBinaryWriter` provides necessary methods for that:
    1. `ValueTask WriteAsync<TArg>(Action<TArg, IBufferWriter<byte>> writer, TArg arg, CancellationToken token = default)` allows to obtain buffer writer allocated by `IAsyncBinaryWriter`. Usually, it is acquired from the pool. Also, you can cache `writer` delegate instance to prevent allocations for each serialization. `IBufferWriter` has many synchronous extension methods for binary serialization located in `DotNext.Buffers` namespace so the migration will be simple.
    1. `ValueTask WriteAsync(ReadOnlyMemory<byte> input, LengthFormat? lengthFormat = null, CancellationToken token = default)` allows to write a raw block of memory. This method gives you full control of memory allocation mechanism.
1. 100 records per partition probably is too many. Log compaction performance depends on the number of log records in the partition. Try to set it to something between 5 and 10
1. Have you tried `InitialPartitionSize` configuration parameter? It allows to pre-allocate partition files on disk and reduces fragmentation of partition files.

Now let's take a look at existing implementation of `PersistentState`:
1. Background compaction is challenging for a few reasons. First of all, the rate of incoming log entries can be very high and compaction will not have enough time to erase unused partitions. In this case, disk space can grow infinitely. Another problem here is lock state: during compaction I need to have read lock on all partitions. Read lock prevent appending new entries. As a result, `AppendEntries` have to wait for compaction.
2. Splitting buffer size doesn't seem to be hard. I can do that. However, it's very strange that buffer size has different performance for partitions and snapshot file. Let's return back to this point when you are done with the first part.
3. Concurrent writes require to change a whole architecture of existing `PersistentState` class. And I'm still not sure that it's possible. Log records stored in partition file sequentially. Without knowledge of the size of previous log record I can't calculate the offset in the file. Moreover, two log record with the same log index cannot be stored simultaneously.

If you have clear ideas of how to do concurrent writes on disk without lock then you can implement `IPersistentState` from scratch or make PR.

**Reply by potrusil-osi** [reply]
Thank you for the suggestions!

1. I will try to optimize serialization as you described. You wrote that it would make sense for records with size less than 1 KB. My records are around 12 KB - would that approach be a problem with this size? And yes, there are many primitive types and unfortunately also many strings that I serialize with UTF8 encoding context.
2. I can try this, but my measurements suggest that the biggest bottleneck during compaction is reading (deserializing) the previous snapshot, and also writing the resulting snapshot into the file. Neither of these will be eliminated if I decrease the number of records per partition. The average snapshot contains around 700 records...
3. Yes, I have played a lot with `InitialPartitionSize`, too. I'm not sure how much it helped, though.

The current implementation:

1. The reasoning makes sense. Maybe one suggestion would be to allow the compaction on demand. Since I understand the state machine I'm implementing (it is a queue with many _inserts_ and occasional _clears_) I know exactly when compaction would result in the actual compaction of the disk space (right after a _clear_ has been committed).
2. OK
3. I agree. I've found some designs of WAL where authors tried to optimize concurrent access (I think it was MySQL), but it was way too complicated. I'll try to solve this problem on the application layer.

**Reply by potrusil-osi** [reply]
I just realized that the "compaction on demand" is already possible through a smart implementation of `PersistentState.CreateSnapshotBuilder`. It won't be on the background but it might result in better overall performance.

**Reply by sakno** [reply]
@potrusil-osi , I can add public method to `PersistentState` for manual compaction. Actually, it's there but not available publicly and called `ForceCompaction`. Also, I can add `AutomaticCompaction` configuration property to options which is **true** by default. You'll able to set it to **false** and call `ForceCompactionAsync` explicitly when needed.

**Reply by potrusil-osi** [reply]
Public `ForceCompaction` and the `AutomaticCompaction` option sound great!

**Reply by sakno** [reply]
You said that snapshot contains ~ 700 records. I think this is not correct strategy to just save committed log entries to the snapshot as-is. Compaction is a process of interpreting and removing operations encoded in log records.
For instance, you have two committed log entries:
```
(1) X := 10
(2) Y := 20
(3) X := 30
```

During snapshot building process, you can remove the first operation because `X` is rewritten by subsequent operation. In your case with the implementation of the queue, you can try to remove the submitted message from the snapshot if all declared consumers have already received messages.

**Reply by sakno** [reply]
`ForceCompactionAsync` is now available in `develop` branch.

**Reply by potrusil-osi** [reply]
I wasn't too specific when I mentioned the queue. It is not the typical queue between publishers and consumers. It is rather a "reliable backing store" for an application which steadily enqueues items into it, and occasionaly clears those that are not needed to be backed up anymore. In my test the queue grows to approximately 700 items and then it is cleared.

Thanks for the quick `ForceCompactionAsync`! I'll definitely have a look.

**Reply by sakno** [reply]
I have a good news. There is a way to do the compaction in the background. The algorithm is based on a few empiric assumptions:
  * Writes causes adding new log entries to the tail of the log. So fully committed partitions can be compacted in parallel with write operations
  * Committed log entries in the beginning of the log are rarely requested for reads (only if a new cluster member is added which requires full replication)

Also I found a way to improve the performance of write operations. At the moment, serialization and disk I/O for each log entry can be done only if write lock is acquired. The actual content of the log entry can be moved out of the log itself. For instance, I can save it to the separated file and then use file name as an actual content of the log entry. Thus, serialization and persistence of the log entry can be done out of the write lock and overhead for each log entry will be predictable and constant regardless of the payload size.

However, all these things are hard to implement and require a lot of investigation.

**Reply by potrusil-osi** [reply]
Hi @sakno, I have experimented a lot and was able to reduce impact of some parts.

Interestingly, similar to what you have suggested in your latest message, I now serialize most of the content of each entry into a buffer before `PersistentState.AppendAsync` is called. This has reduced time of AppendAsync to half - it is down to 20ms (from 40ms) in my tests (with 30 client threads).

Another optimization is skipping deserialization of those entries (most of their content) that I know have been cleared by the application (this is specific to my state machine that I described earlier). This helped reduce the compaction time down to about 35ms (from 100ms). I believe it could be even better if I was able to skip the unused memory blocks without copying them into a buffer first (`IAsyncBinaryReader` allows me to only read a block of memory into a buffer).

But I got the best performance when I used the new `ForceCompactionAsync` to run at certain times on the background. Well, I actually changed the method on my machine a bit - instead of locking the whole compaction, I lock only those parts that need it:
* The 1st part - finding the partition
* The 4th part - persisting the snapshot and
* Each iteration of the 5th part - removing partitions

I like your ideas. It sounds you'd be able to improve everything even more. Hopefully you'll have time to work on those.

**Reply by sakno** [reply]
I can add skipping functionality to `IAsyncBinaryReader`, it's not a problem. How do you like to see this method?
```csharp
ValueTask SkipAsync(long length, CancellationToken token);
```

Is it OK?

**Reply by potrusil-osi** [reply]
Yes, that is exactly what I meant. Thanks!

**Reply by sakno** [reply]
You can check `develop` branch for this functionality. Unfortunately, I can't tell you ETA of release at this moment.

**Reply by potrusil-osi** [reply]
I think there is a bug in `SkipAsync` when `StreamSegment` is the stream used by the reader - updating its `Position` doesn't update the position of the underlying stream.

**Reply by potrusil-osi** [reply]
Or maybe I'm wrong where the problem is. But I'm getting `EndOfStreamException` when an entry is being read after the previous one has been skipped.

**Reply by potrusil-osi** [reply]
It is either `EndOfStreamException` or `OutOfMemoryException`. Both appear to be the result of a stream corruption when the entry is being read/deserialized.

**Reply by sakno** [reply]
I would suggest to move further discussions to #57. I'll check the problem with `SkipAsync` method.
